### ISSUE TITLE
Reduce substantial RAM consumption 

### ISSUE THREAD

**Issue by Seb35** [ProfessionalWiki/SCSS#4]
This extension, when used with the extension Bootstrap, uses a lot of RAM memory, too much for the default `memory_limit` of php-fpm (128M), and, as a result, the styles of Bootstrap are not computed (HTTP error 500 in load.php) and the page is mainly unstyled (some styles from other extensions are used).

I have no immediate solution to solve it, even increasing `memory_limit` to 512M is not enough (and it would be a big value for production). The only hack I found is to keep enabled `$egScssCacheType` to some cache, execute the following code with `php maintenance/eval.php` (there is no memory limit in CLI), re-execute it before the cache is purged and/or add some feedback loop to recompute it when the cache is lost.
```php
$rl = MediaWiki\MediaWikiServices::getInstance()->getResourceLoader();
$request = new FauxRequest( [ 'useskin' => 'chameleon', 'uselang' => 'fr', 'skin' => 'chameleon', 'lang' => 'fr' ] );
$ctx = new ResourceLoaderContext( $rl, $request );
$rl->makeModuleResponse( $ctx, [ $rl->getModule( 'zzz.ext.bootstrap.styles' ) ] );
```

**Comment by s7eph4n** [comment]
There exists a php module for SCSS: https://sass-lang.com/libsass#php

Apart from the obvious drawback that it is a php module that is not available out of the box on most servers and would need installing, this module is probably much faster and much less resource hungry then the pure PHP SCSS compiler. Might be worth offering it as an optional compiler.

**Comment by kghbln** [comment]
I am not sure if [absalomedia/sassphp](https://github.com/absalomedia/sassphp) was meant when referring to a php module for scss. I failed installing it however I only allocated 15 mins for it. I suspect that I will not be the only one with issues here. Anyhow I also realized a pretty poor performance for Chameleon in connection with this extension, i.e. with both of them invoked. Not sure if SCSS is a hard required, at least the docu for Chameleon does not say so.

**Comment by s7eph4n** [comment]
Yes, it is. Because the Bootstrap styles need compiling. Otherwise it would not be possible to customize them. However, this compilation is only done once, then the styles are cached.

**Comment by kghbln** [comment]
Thanks for clarifying. Will add this to the docu. It may be that it was just the first call to Chameleon that was slow. Only updated to 2.x last week so not much experience here yet.

**Comment by JeroenDeDauw** [comment]
We switched to a more recent version of the used library with https://github.com/ProfessionalWiki/SCSS/commit/39f20d4736fef8ee0bfaae99b49db39bf12b220b. Perhaps this brings some performance improvements.

**Comment by kghbln** [comment]
> Will add this to the docu.

I guess #7 will save me from doing so. :)

**Comment by JeroenDeDauw** [comment]
Turning this into a library does not affect caching or the need to compile styles.

**Comment by kghbln** [comment]
> Turning this into a library does not affect caching or the need to compile styles.

No but I do not have to document everywhere the SCSS extension is a hard requirement for Chameleon. ;)

**Comment by Seb35** [comment]
Sorry about the delay, we had previously other issues with chameleon/Bootstrap/composer whose I tried to better understand if it was on our infrastructure or in the code, and it turns these other issues were finally solved by #7 and MediaWiki autoloading & requires mechanism by @cicalese (249fd49 in Bootstrap and df6a777 + 43fcdae in chameleon).

About this specific issue, this is now solved on our infrastructure, probably thanks to c5af4d4a5e3cb - styles are correctly computed with `memory_limit = 128M`. Consequently I close the issue. Thanks!
